contrib/cirrus/lib.sh

```diff
@@ -4,127 +4,124 @@
-if [[ -r "$HOME/$ENVLIB" ]] && [[ -n "$CI" ]]
+# Defaults when not running under CI
+export CI="${CI:-false}"
+CIRRUS_CI="${CIRRUS_CI:-false}"
+CONTINUOUS_INTEGRATION="${CONTINUOUS_INTEGRATION:-false}"
+CIRRUS_REPO_NAME=${CIRRUS_REPO_NAME:-libpod}
+CIRRUS_BASE_SHA=${CIRRUS_BASE_SHA:-unknown$(date +%s)}  # difficult to reliably discover
+CIRRUS_BUILD_ID=${CIRRUS_BUILD_ID:-$RANDOM$(date +%s)}  # must be short and unique
+# Vars. for image-building
+PACKER_VER="1.3.5"
+# CSV of cache-image names to build (see $PACKER_BASE/libpod_images.json)
+PACKER_BUILDS="${PACKER_BUILDS:-ubuntu-18,fedora-29,fedora-28,rhel-7,centos-7}"
+# Google-maintained base-image names
+UBUNTU_BASE_IMAGE="ubuntu-1804-bionic-v20181203a"
+CENTOS_BASE_IMAGE="centos-7-v20181113"
+# Manually produced base-image names (see $SCRIPT_BASE/README.md)
+FEDORA_BASE_IMAGE="fedora-cloud-base-29-1-2-1541789245"
+PRIOR_FEDORA_BASE_IMAGE="fedora-cloud-base-28-1-1-1544474897"
+FAH_BASE_IMAGE="fedora-atomichost-29-20181025-1-1541787861"
+# RHEL image must be imported, native image bills for subscription.
+RHEL_BASE_IMAGE="rhel-guest-image-7-6-210-x86-64-qcow2-1548099756"
+BUILT_IMAGE_SUFFIX="${BUILT_IMAGE_SUFFIX:--$CIRRUS_REPO_NAME-${CIRRUS_BUILD_ID}}"
+RHSM_COMMAND="${RHSM_COMMAND:-/bin/true}"
+
+# Safe env. vars. to transfer from root -> $ROOTLESS_USER  (go env handled separetly)
+ROOTLESS_ENV_RE='(CIRRUS_.+)|(ROOTLESS_.+)|(.+_IMAGE.*)|(.+_BASE)|(.*DIRPATH)|(.*FILEPATH)|(SOURCE.*)|(DEPEND.*)|(.+_DEPS_.+)|(OS_REL.*)|(.+_ENV_RE)|(TRAVIS)|(CI.+)'
+# Unsafe env. vars for display
+SECRET_ENV_RE='(IRCID)|(RHSM)|(ACCOUNT)|(^GC[EP]..+)|(SSH)'
+
+SPECIALMODE="${SPECIALMODE:-none}"
+export CONTAINER_RUNTIME=${CONTAINER_RUNTIME:-podman}
+# When running as root, this may be empty or not, as a user, it MUST be set.
+if [[ "$USER" == "root" ]]
 then
-    # Make sure this is always loaded
-    source "$HOME/$ENVLIB"
+    ROOTLESS_USER="${ROOTLESS_USER:-}"
+else
+    ROOTLESS_USER="${ROOTLESS_USER:-$USER}"
 fi
 
-# Pass in a line delimited list of, space delimited name/value pairs
-# exit non-zero with helpful error message if any value is empty
+# GCE image-name compatible string representation of distribution name
+OS_RELEASE_ID="$(egrep -m 1 '^ID=' /etc/os-release | cut -d = -f 2 | tr -d \' | tr -d \")"
+# GCE image-name compatible string representation of distribution major version
+OS_RELEASE_VER="$(egrep -m 1 '^VERSION_ID=' /etc/os-release | cut -d = -f 2 | tr -d \' | tr -d \" | cut -d '.' -f 1)"
+# Combined to ease soe usage
+OS_REL_VER="${OS_RELEASE_ID}-${OS_RELEASE_VER}"
+
+# Allow checking of required, dependent variables w/ better debugging than 'set -u'
+# Pass in a label (like "${FUNCNAME[0]}()" ) and a line separated list of,
+# space delimited name/value pairs.  Exits non-zero with helpful error
+# message if any value is empty.
 req_env_var() {
-    echo "$1" | while read NAME VALUE
+    [[ -n "$1" ]] || die 18 "${FUNCNAME[0]}() expects first argument to be non-empty"
```

Comment:
There are an awful lot of complicated scary FUNCNAMEs when invoking this function. Did you consider:
```bash
    caller="$1"
    if [ -n "$1" -o "$1" = "-" ]; then caller=${FUNCNAME[1]};fi
    ...
    req_env_var - "etc etc
```

Reply:
Actually there's a lot more unnecessary duplication: every invocation of req_env_var is of the form `... FOO $FOO`. Much better to eliminate the duplication:
```bash
req_env_var() {
    caller="$1";shift
    if [ -n "$caller" -o "$caller" = "-" ]; then caller=${FUNCNAME[1]};fi

    for i;do
        if [ -z "${!i}" ]; then die 9 "$caller requires \$$i to be non-empty";fi
    done
}
...
    req_env_var - FOO BAR FUBAR GOPATH PATH ...
```

Reply:
Don't look too closely, you'll likely find plenty of other similar duct-tape :smile: 

I like your replacement implementation.  For functions calling 'req_env_var' do I need to change them to pass a literal "-" or is that (somehow) automatic?

Reply:
It would need to be explicit, in all instances that are currently FUNCNAME (but not the `basename $0` ones).

I didn't actually look closely at the `basename` ones; if those are inline (not in a function), it would be even cleaner to get rid of the `-` thing completely, and in the function itself, if FUNCNAME[1] is empty, default to basename $0. But again, I haven't looked closely.

Reply:
kk, running with the "-" for now.

Reply:
actually...yeah, checking if `FUNCNAME[1]` is empty, is much cleaner.